synapseclient/client.py

```diff
@@ -765,131 +769,108 @@ def _getWithEntityBundle(self, entityBundle, entity=None, **kwargs):
-                    entity['synapseStore'] = False
 
-        return entity
+    def _download_file_entity(self, downloadLocation, entity, ifcollision, submission):
+        # set the initial local state
+        entity.path = None
+        entity.files = []
+        entity.cacheDir = None
+
+        # check to see if an UNMODIFIED version of the file (since it was last downloaded) already exists
+        # this location could be either in .synapseCache
+        # or a user specified location to which the user previously downloaded the file
+        cached_file_path = self.cache.get(entity.dataFileHandleId, downloadLocation)
+
+        #location in .synapseCache where the file would be corresponding to its FileHandleId
+        synapseCache_location = self.cache.get_cache_dir(entity.dataFileHandleId)
+
+        file_name = entity.name if cached_file_path is None else os.path.basename(cached_file_path)
+
+        #Decide the best download location for the file
+        if downloadLocation is not None:
+            # Make sure the specified download location is a fully resolved directory
+            downloadLocation = os.path.expandvars(os.path.expanduser(downloadLocation))
+            if os.path.isfile(downloadLocation):
+                raise ValueError("Parameter 'downloadLocation' should be a directory, not a file.")
+        elif cached_file_path is not None:
+            #file already cached so use that as the download location
+            downloadLocation = os.path.dirname(cached_file_path)
+        else:
+            #file not cached and no user-specified location so default to .synapseCache
+            downloadLocation = synapseCache_location
+
+        downloadPath = self._resolve_download_path(downloadLocation, file_name, ifcollision, synapseCache_location, cached_file_path)
+        if downloadPath is None:
+            return
+
+        if cached_file_path is not None: #copy from cache
+            if downloadPath != cached_file_path:
+                # create the foider if it does not exist already
+                if not os.path.exists(downloadLocation):
+                    os.makedirs(downloadLocation)
+                shutil.copy(cached_file_path, downloadPath)
+
+        else: #download the file from URL (could be a local file)
+            objectType = 'FileEntity' if submission is None else 'SubmissionAttachment'
+            objectId = entity['id'] if submission is None else submission
+            fileResult = self._getFileHandleDownload(entity.dataFileHandleId,
+                                                     objectId, objectType)
+
+            # reassign downloadPath because if url points to local file (e.g. file://~/someLocalFile.txt)
+            # it won't be "downloaded" and, instead, downloadPath will just point to '~/someLocalFile.txt'
+            downloadPath = self._downloadFileHandle(fileResult['preSignedURL'],
```

Comment:
Modify _downloadFileHandle to copy from the specifiedl local file path to downloadPath